### DISCUSSION TITLE
What could be the possible causes of changing machine id?

### DISCUSSION THREAD

**Discussion by SethXie** [dashingsoft/pyarmor#1373] [Q&A]
I had an activated Group License on my machine and was able to run `pyarmor gen` commend previously. But recently I got this:
`ERROR    request license token faild (3)` when I used the same `pyarmor gen` commend, and the license type became **trial** after the error:

```
$ pyarmor --version
Pyarmor 8.2.9 (trial), 000000, non-profits

License Type    : pyarmor-trial
License No.     : pyarmor-vax-000000
License To      :
License Product : non-profits

BCC Mode        : No
RFT Mode        : No

Notes
* Can't obfuscate big script and mix str
``` 
I wanted to re-register the device but failed:
```
$ pyarmor reg pyarmor-device-regfile-6012.1.zip
INFO     Python 3.8.10
INFO     Pyarmor 8.2.0 (trial), 000000, non-profits
INFO     Platform linux.x86_64
INFO     register "pyarmor-device-regfile-6012.1.zip"
ERROR    this regfile is not for this device
(env) FAIL
```
I run the below python function to get the hd info, and noticed the result is different from my `machine_id` in `.pyarmor/group/pyarmor-group-device.1`
```
>>> from pyarmor.cli.core import Pytransform3
>>> Pytransform3.get_hd_info(11)
```

I'm sure the hard-drive and mac address haven't been changed for my device. The same issue happened multiple times. We're using linux Virtual Machines - I'm not sure if rebooting it will impact the hd_info.

What's the logic of `get_hd_info`? What are the possible factors that lead to the changing machine ID?

**Comment by jondy** [comment]
>INFO     Pyarmor 8.2.0 (trial), 000000, non-profits
>INFO     Platform linux.x86_64

Please upgrade pyarmor to 8.2.9 in all devices, it changes the logic of machine id to support unlimited docker containers.

**Reply by SethXie** [reply]
@jondy Yes I noticed the change from `Pytransform3.get_hd_info(10)` to `Pytransform3.get_hd_info(11)` also resulted in different machine ids for the same device.  I would like to fix the license for the same device. Do you mind tell me what's the meaning of this parameter? Will the logic be changed often with upgrading pyarmor version?

**Comment by jondy** [comment]
It will not changed often.

**Comment by jondy** [comment]
And it's compatible backward, that is to say, the device file generated by old pyarmor is recognized by latest pyarmor, but the old pyarmor doesn't know the new one.
